diff --git a/src/core/lombok/AllArgsConstructor.java b/src/core/lombok/AllArgsConstructor.java
index f0e0e3f..b6179c9 100644
--- a/src/core/lombok/AllArgsConstructor.java
+++ b/src/core/lombok/AllArgsConstructor.java
@@ -44,6 +44,8 @@ public @interface AllArgsConstructor {
 	 */
 	String staticName() default "";
 	
+	boolean callSuper() default false;
+	
 	/**
 	 * Sets the access level of the constructor. By default, generated constructors are {@code public}.
 	 */
diff --git a/src/core/lombok/Data.java b/src/core/lombok/Data.java
index 488de64..61e13c6 100644
--- a/src/core/lombok/Data.java
+++ b/src/core/lombok/Data.java
@@ -45,6 +45,7 @@ import java.lang.annotation.Target;
 @Target(ElementType.TYPE)
 @Retention(RetentionPolicy.SOURCE)
 public @interface Data {
+	
 	/**
 	 * If you specify a static constructor name, then the generated constructor will be private, and
 	 * instead a static factory method is created that other classes can use to create instances.
@@ -57,4 +58,12 @@ public @interface Data {
 	 * Default: No static constructor, instead the normal constructor is public.
 	 */
 	String staticConstructor() default "";
+	
+	/**
+	 * Include the result of the superclass's implementation of {@code toString} in the output.
+	 * And Call on the superclass's implementations of {@code equals} and {@code hashCode} before calculating
+	 * for the fields in this class.
+	 * <strong>default: false</strong>
+	 */
+	boolean callSuper() default false;
 }
diff --git a/src/core/lombok/NoArgsConstructor.java b/src/core/lombok/NoArgsConstructor.java
index 1a885b6..42059ea 100644
--- a/src/core/lombok/NoArgsConstructor.java
+++ b/src/core/lombok/NoArgsConstructor.java
@@ -47,6 +47,8 @@ public @interface NoArgsConstructor {
 	 */
 	String staticName() default "";
 	
+	boolean callSuper() default false;
+	
 	/**
 	 * Sets the access level of the constructor. By default, generated constructors are {@code public}.
 	 */
diff --git a/src/core/lombok/RequiredArgsConstructor.java b/src/core/lombok/RequiredArgsConstructor.java
index ba56726..9042fe8 100644
--- a/src/core/lombok/RequiredArgsConstructor.java
+++ b/src/core/lombok/RequiredArgsConstructor.java
@@ -44,6 +44,8 @@ public @interface RequiredArgsConstructor {
 	 */
 	String staticName() default "";
 	
+	boolean callSuper() default false;
+	
 	/**
 	 * Sets the access level of the constructor. By default, generated constructors are {@code public}.
 	 */
diff --git a/src/core/lombok/eclipse/handlers/HandleData.java b/src/core/lombok/eclipse/handlers/HandleData.java
index 71027c0..2a44cf4 100644
--- a/src/core/lombok/eclipse/handlers/HandleData.java
+++ b/src/core/lombok/eclipse/handlers/HandleData.java
@@ -58,10 +57,15 @@ public class HandleData extends EclipseAnnotationHandler<Data> {
 		//for whatever reason, though you can find callers of that one by focusing on the class name itself
 		//and hitting 'find callers'.
 		
+		Data data = annotation.getInstance();
+		String staticConstructorName = data.staticConstructor();
+		Boolean callSuper = data.callSuper();
+		if (!annotation.isExplicit("callSuper")) callSuper = null;
+		
 		new HandleGetter().generateGetterForType(typeNode, annotationNode, AccessLevel.PUBLIC, true);
 		new HandleSetter().generateSetterForType(typeNode, annotationNode, AccessLevel.PUBLIC, true);
-		new HandleEqualsAndHashCode().generateEqualsAndHashCodeForType(typeNode, annotationNode);
-		new HandleToString().generateToStringForType(typeNode, annotationNode);
-		new HandleConstructor().generateRequiredArgsConstructor(typeNode, AccessLevel.PUBLIC, ann.staticConstructor(), true, ast);
+		new HandleEqualsAndHashCode().generateEqualsAndHashCodeForType(typeNode, annotationNode, callSuper);
+		new HandleToString().generateToStringForType(typeNode, annotationNode, callSuper);
+		new HandleConstructor().generateRequiredArgsConstructor(typeNode, AccessLevel.PUBLIC, staticConstructorName, true, ast);
 	}
 }
diff --git a/src/core/lombok/eclipse/handlers/HandleEqualsAndHashCode.java b/src/core/lombok/eclipse/handlers/HandleEqualsAndHashCode.java
index bbb1d56..3227fc1 100644
--- a/src/core/lombok/eclipse/handlers/HandleEqualsAndHashCode.java
+++ b/src/core/lombok/eclipse/handlers/HandleEqualsAndHashCode.java
@@ -103,7 +103,7 @@ public class HandleEqualsAndHashCode extends EclipseAnnotationHandler<EqualsAndH
 		}
 	}
 	
-	public void generateEqualsAndHashCodeForType(EclipseNode typeNode, EclipseNode errorNode) {
+	public void generateEqualsAndHashCodeForType(EclipseNode typeNode, EclipseNode errorNode, Boolean callSuper) {
 		for (EclipseNode child : typeNode.down()) {
 			if (child.getKind() == Kind.ANNOTATION) {
 				if (Eclipse.annotationTypeMatches(EqualsAndHashCode.class, child)) {
@@ -113,7 +113,7 @@ public class HandleEqualsAndHashCode extends EclipseAnnotationHandler<EqualsAndH
 			}
 		}
 		
-		generateMethods(typeNode, errorNode, null, null, null, false, FieldAccess.GETTER);
+		generateMethods(typeNode, errorNode, null, null, callSuper, false, FieldAccess.GETTER);
 	}
 	
 	@Override public void handle(AnnotationValues<EqualsAndHashCode> annotation,
diff --git a/src/core/lombok/eclipse/handlers/HandleToString.java b/src/core/lombok/eclipse/handlers/HandleToString.java
index f167916..9be572e 100644
--- a/src/core/lombok/eclipse/handlers/HandleToString.java
+++ b/src/core/lombok/eclipse/handlers/HandleToString.java
@@ -81,7 +81,7 @@ public class HandleToString extends EclipseAnnotationHandler<ToString> {
 		}
 	}
 	
-	public void generateToStringForType(EclipseNode typeNode, EclipseNode errorNode) {
+	public void generateToStringForType(EclipseNode typeNode, EclipseNode errorNode, Boolean callSuper) {
 		for (EclipseNode child : typeNode.down()) {
 			if (child.getKind() == Kind.ANNOTATION) {
 				if (Eclipse.annotationTypeMatches(ToString.class, child)) {
@@ -95,7 +95,7 @@ public class HandleToString extends EclipseAnnotationHandler<ToString> {
 		try {
 			includeFieldNames = ((Boolean)ToString.class.getMethod("includeFieldNames").getDefaultValue()).booleanValue();
 		} catch (Exception ignore) {}
-		generateToString(typeNode, errorNode, null, null, includeFieldNames, null, false, FieldAccess.GETTER);
+		generateToString(typeNode, errorNode, null, null, includeFieldNames, callSuper, false, FieldAccess.GETTER);
 	}
 	
 	public void handle(AnnotationValues<ToString> annotation, Annotation ast, EclipseNode annotationNode) {
@@ -286,7 +286,7 @@ public class HandleToString extends EclipseAnnotationHandler<ToString> {
 	private String getTypeName(EclipseNode type) {
 		String typeName = getSingleTypeName(type);
 		EclipseNode upType = type.up();
-		while (upType.getKind() == Kind.TYPE) {
+		if (upType != null) while (upType.getKind() == Kind.TYPE) {
 			typeName = getSingleTypeName(upType) + "." + typeName;
 			upType = upType.up();
 		}
diff --git a/src/core/lombok/javac/handlers/HandleData.java b/src/core/lombok/javac/handlers/HandleData.java
index 79fdf76..619ef91 100644
--- a/src/core/lombok/javac/handlers/HandleData.java
+++ b/src/core/lombok/javac/handlers/HandleData.java
@@ -52,13 +52,16 @@ public class HandleData extends JavacAnnotationHandler<Data> {
 			return;
 		}
 		
-		String staticConstructorName = annotation.getInstance().staticConstructor();
+		Data data = annotation.getInstance();
+		String staticConstructorName = data.staticConstructor();
+		Boolean callSuper = data.callSuper();
+		if (!annotation.isExplicit("callSuper")) callSuper = null;
 		
 		// TODO move this to the end OR move it to the top in eclipse.
 		new HandleConstructor().generateRequiredArgsConstructor(typeNode, AccessLevel.PUBLIC, staticConstructorName, true, annotationNode);
 		new HandleGetter().generateGetterForType(typeNode, annotationNode, AccessLevel.PUBLIC, true);
 		new HandleSetter().generateSetterForType(typeNode, annotationNode, AccessLevel.PUBLIC, true);
-		new HandleEqualsAndHashCode().generateEqualsAndHashCodeForType(typeNode, annotationNode);
-		new HandleToString().generateToStringForType(typeNode, annotationNode);
+		new HandleEqualsAndHashCode().generateEqualsAndHashCodeForType(typeNode, annotationNode, callSuper);
+		new HandleToString().generateToStringForType(typeNode, annotationNode, callSuper);
 	}
 }
diff --git a/src/core/lombok/javac/handlers/HandleEqualsAndHashCode.java b/src/core/lombok/javac/handlers/HandleEqualsAndHashCode.java
index afd299e..8af6f9d 100644
--- a/src/core/lombok/javac/handlers/HandleEqualsAndHashCode.java
+++ b/src/core/lombok/javac/handlers/HandleEqualsAndHashCode.java
@@ -100,7 +100,7 @@ public class HandleEqualsAndHashCode extends JavacAnnotationHandler<EqualsAndHas
 		generateMethods(typeNode, annotationNode, excludes, includes, callSuper, true, fieldAccess);
 	}
 	
-	public void generateEqualsAndHashCodeForType(JavacNode typeNode, JavacNode source) {
+	public void generateEqualsAndHashCodeForType(JavacNode typeNode, JavacNode source, Boolean callSuper) {
 		for (JavacNode child : typeNode.down()) {
 			if (child.getKind() == Kind.ANNOTATION) {
 				if (Javac.annotationTypeMatches(EqualsAndHashCode.class, child)) {
@@ -110,7 +110,7 @@ public class HandleEqualsAndHashCode extends JavacAnnotationHandler<EqualsAndHas
 			}
 		}
 		
-		generateMethods(typeNode, source, null, null, null, false, FieldAccess.GETTER);
+		generateMethods(typeNode, source, null, null, callSuper, false, FieldAccess.GETTER);
 	}
 	
 	private void generateMethods(JavacNode typeNode, JavacNode source, List<String> excludes, List<String> includes,
diff --git a/src/core/lombok/javac/handlers/HandleToString.java b/src/core/lombok/javac/handlers/HandleToString.java
index 06e70e7..3c53a12 100644
--- a/src/core/lombok/javac/handlers/HandleToString.java
+++ b/src/core/lombok/javac/handlers/HandleToString.java
@@ -95,7 +95,7 @@ public class HandleToString extends JavacAnnotationHandler<ToString> {
 		generateToString(typeNode, annotationNode, excludes, includes, ann.includeFieldNames(), callSuper, true, fieldAccess);
 	}
 	
-	public void generateToStringForType(JavacNode typeNode, JavacNode errorNode) {
+	public void generateToStringForType(JavacNode typeNode, JavacNode errorNode, Boolean callSuper) {
 		for (JavacNode child : typeNode.down()) {
 			if (child.getKind() == Kind.ANNOTATION) {
 				if (Javac.annotationTypeMatches(ToString.class, child)) {
@@ -109,7 +109,7 @@ public class HandleToString extends JavacAnnotationHandler<ToString> {
 		try {
 			includeFieldNames = ((Boolean)ToString.class.getMethod("includeFieldNames").getDefaultValue()).booleanValue();
 		} catch (Exception ignore) {}
-		generateToString(typeNode, errorNode, null, null, includeFieldNames, null, false, FieldAccess.GETTER);
+		generateToString(typeNode, errorNode, null, null, includeFieldNames, callSuper, false, FieldAccess.GETTER);
 	}
 	
 	private void generateToString(JavacNode typeNode, JavacNode source, List<String> excludes, List<String> includes,
